src/dag-node.js

```diff
@@ -1,281 +1,83 @@
-          // 2. create new node
-        } else if (typeof nameOrLink === 'string') {
-          // It's a name
-          const name = nameOrLink
-          if ((nodeOrMultihash.constructor &&
-             nodeOrMultihash.constructor.name === 'DAGNode')) {
-            // It's a node
-            // TODO
-            // 1. create link from name + node
-            // 2. add to the set of links
-            // 3. create new node
-          } else {
-            // It's a multihash
-            let multihash = nodeOrMultihash
-            if (typeof multihash === 'string') {
-              multihash = mh.fromB58String(multihash)
-            }
-            // TODO
-            // 1. create link from name + multihash
-            // 2. add to the set of links
-            // 3. create new node
-          }
-        }
-        callback(new Error('invalid arguments'))
-      },
-      rm: (nameOrMultihash, callback) => {
-        let links
-        if (typeof nameOrMultihash === 'string') {
-          const name = nameOrMultihash
-          links = this.links.filter((link) => {
-            if (link.name === name) {
-              return false
-            } else {
-              return true
-            }
-          })
-        } else if (Buffer.isBuffer(nameOrMultihash)) {
-          const multihash = nameOrMultihash
-          links = this.links.filter((link) => {
-            if (link.hash.equals(multihash)) {
-              return false
-            } else {
-              return true
-            }
-          })
-        } else {
-          callback(new Error('first arg needs to be a name or multihash'))
-        }
-
-        let dataClone
-
-        if (this.data && this.data.length > 0) {
-          dataClone = new Buffer(this.data.length)
-          this.data.copy(dataClone)
-        }
-
-        DAGNode.create(dataClone, links, callback)
-      }
-    }
-  }
-
-  /*
-   * makeLink returns a DAGLink node from a DAGNode
-   */
-  // TODO: this would make more sense as an utility
-  makeLink (node, callback) {
-    node.multihash((err, multihash) => {
-      if (err) {
-        return callback(err)
-      }
-      node.size((err, size) => {
-        if (err) {
-          return callback(err)
-        }
-        callback(null, new DAGLink(null, size, multihash))
-      })
-    })
+    this._data = data || new Buffer(0)
+    this._links = links || []
+    this._serialized = serialized || new Buffer(0) // TODO: default serialized object
+    this._multihash = multihash || new Buffer(0) // TODO: default multihash object
   }
 
-  /*
-   * clone - returns a clone of the DAGNode
-   */
-  clone (callback) {
-    let dataClone
-    let linksClone
-
-    if (this.data && this.data.length > 0) {
-      dataClone = new Buffer(this.data.length)
-      this.data.copy(dataClone)
+  toJSON() {
+    return {
+      data: this.data,
+      links: this.links.map((l) => l.json),
+      hash: mh.toB58String(this.multihash),
+      size: this.size
```

Comment:
You're right. Same for the other toJSON() props. Will change it so that they all get calculated in the constructor. A small speed improvement is a big speed improvement :)